compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -492,23 +492,45 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
           case AndType(tp21, tp22) => constrainRHSVars(tp21) && constrainRHSVars(tp22)
           case _ => true
 
-        widenOK
-        || joinOK
-        || (tp1.isSoft || constrainRHSVars(tp2)) && recur(tp11, tp2) && recur(tp12, tp2)
-        || containsAnd(tp1)
-            && !joined
-            && {
-              joined = true
-              try inFrozenGadt(recur(tp1.join, tp2))
-              finally joined = false
-            }
-            // An & on the left side loses information. We compensate by also trying the join.
-            // This is less ad-hoc than it looks since we produce joins in type inference,
-            // and then need to check that they are indeed supertypes of the original types
-            // under -Ycheck. Test case is i7965.scala.
-            // On the other hand, we could get a combinatorial explosion by applying such joins
-            // recursively, so we do it only once. See i14870.scala as a test case, which would
-            // loop for a very long time without the recursion brake.
+        def hardenTypeVars(tp2: Type): Unit = tp2.dealiasKeepRefiningAnnots match
+          case tvar: TypeVar if constraint.contains(tvar.origin) =>
+            tvar.widenUnions = false
```

Comment:
This seems very dangerous to me. The particular subtype check we're doing might be part of a TyperState that ends up being rolled back, but the mutation here will be preserved. This will make it extremely hard to debug any issue involving widening, and any change in inference could break existing code because this flag happens to be toggled.

Reply:
[bccbaf7](https://github.com/lampepfl/dotty/pull/15632/commits/bccbaf7d07ed237da2215bffa54081536072f932)
now keeps this info in TyperState.